File: StringReduction.java
import java.util.*;

class StringReduction {
    static String reduce(String str) {
        HashMap<String, String> replacements = new HashMap<String, String>();
        replacements.put("ab", "c");
        replacements.put("ac", "b");
        replacements.put("bc", "a");
        replacements.put("ba", "c");
        replacements.put("ca", "b");
        replacements.put("cb", "a");

        String lastString = "";
        while (str != lastString) {
            lastString = str;
            for (String i : replacements.keySet()) {
                str = str.replace(i, replacements.get(i));
            }
        }
        String s = String.valueOf(str.length());
        return s;
    }
    public static void main(String[] args) {
        String str = "cab";
        System.out.println(reduce(str));
    }
}

File: StringReduction.js
const replacements = {
    "ab" : "c",
    "ac" : "b",
    "bc" : "a",
    "ba" : "c",
    "ca" : "b",
    "cb" : "a"
};

function strReduce(str) {
    let lastString = '';
    while (str != lastString) {
        lastString = str;
        for (char in replacements) {
            str = str.replace(char, replacements[char]);
        }
    }
    return str.length;
}

console.log(strReduce("bcab")); // Should equal 1
console.log(strReduce("cab")); // Should equal 2
console.log(strReduce("abcabc")); // Should equal 2
console.log(strReduce("cccc")); // Should equal 4
console.log(strReduce("baccabcbc")); // Should equal 1